### PR TITLE
PRT-236 Accept db generics

### PR DESCRIPTION
**Pull request by rpanasiuk**
First of all, thanks for your contribution! :-)

Please makes sure these boxes are checked before submitting your PR, thank you!

* [ ] Make sure you propose PR to correct branch: hotfix for `master`, feature for latest active branch `feature/x`.
* [ ] Run `npm run lint` and fix those errors before submitting in order to keep consistent code style.
* [ ] Rebase before creating a PR to keep commit history clear.
* [ ] Add some descriptions and refer relative issues for you PR.

### CODE REVIEW

File: packages/orm/src/types.ts [hunk 1]
```diff
@@ -11,6 +11,155 @@ import {
   UPDATE,
 } from "./constants";
 import Table from "./db/Table";
+import { AnyModel, ModelClassMap } from "./Model";
```
Comment (rpanasiuk) [inline packages/orm/src/types.ts:14]: I am planning to do the prettifying as the last step of my work on this fork.

File: packages/orm/src/types.ts [hunk 2]
```diff
@@ -11,6 +11,155 @@ import {
...
+  ? FirstConstructorParam
+  : never
+  : never;
+
+/**
+ * Extends the plan JS object interface, representing the database entry, with {@link ModelField}
+ * 
+ * Mainly used in functions that manipulate db entries.
+ */
+export type RefWithFields<M extends AnyModel> = {
```
Comment (rpanasiuk) [inline packages/orm/src/types.ts:126]: This type is needed as we can be making db entries manipulations using not only serializable data but the model instances / nullish values as well.

File: packages/orm/src/types.ts [hunk 3]
```diff
@@ -11,6 +11,155 @@ import {
...
+    ? never
+    : ExcludeUndefined<MFieldMap[K]> extends AnyModel
+      ? ModelId | undefined
+      : MFieldMap[K];
+};
+
+/**
+ * Excludes undefined type from an union of types.
+ */
+type ExcludeUndefined<T> = Exclude<T, undefined>;
```
Comment (rpanasiuk) [inline packages/orm/src/types.ts:150]: Excluding undefined from the union of types for types comparison (if X extends Y...). Maybe I could have get rid of this helper but IMO having this type does not harm the typing in some cases.

File: packages/orm/src/Model.ts [hunk 4]
```diff
@@ -58,33 +59,33 @@ type ModelFields = {
...
   static isSetUp: boolean;
-  static _session: Session;
-  _fields: ModelData;
-  static reducer: (
-    action: ReduxAction,
-    modelClass: typeof Model,
-    session: Session
+  static _session: SessionWithBoundModels<any>;
+  _fields: ModelRefLike<MFieldMap>;
+  static reducer: <Schema extends AnySchema, ModelClassType extends Schema[keyof Schema]>(
```
[the remark this answers is not in the document]
Comment (mykulyak) [reply packages/orm/src/Model.ts:72]: We can resign from reducers altogether. In 201 they turned out to be too restrictive to use (e.g. when reduced needed to modify multiple models).

File: packages/orm/src/Model.ts [hunk 5]
```diff
@@ -252,27 +253,28 @@ export default class Model {
...
+      const fields = this.fields;
+      const field = fields[key];
       const valuePassed = userProps.hasOwnProperty(key);
       if (!(field instanceof ManyToMany)) {
         if (valuePassed) {
-          const value = userProps[key];
-          props[key] = normalizeEntity(value);
-        } else if (field.getDefault) {
-          props[key] = field.getDefault();
+          const value = (userProps[key as keyof RefWithFields<InstanceType<MClassType>>] as unknown) as AnyModel;
```
Comment (mykulyak) [inline packages/orm/src/Model.ts:269]: I wrote castTo<DestT>(arg) function exactly for this kind of situations - to simplify type casts. You may use it.

File: packages/orm/src/ORM.ts [hunk 6]
```diff
@@ -138,36 +141,36 @@ export default class ORM {
...
     this._setupModelPrototypes(this.registry);
     this._setupModelPrototypes(this.implicitThroughModels);
-    return this.registry.concat(this.implicitThroughModels);
+    return this.registry.concat(this.implicitThroughModels as Schema[keyof Schema][]);
   }
 
   generateSchemaSpec() {
     const models = this.getModelClasses();
-    const tables = models.reduce<Record<string, TableSpec>>((spec, modelClass) => {
+    const tables = models.reduce<{ [K in keyof Schema]: Schema[K] }>((spec, modelClass) => {
```
[the remark this answers is not in the document]
Comment (rpanasiuk) [reply packages/orm/src/ORM.ts]: f1d7b82

**Pull request merged**